crates/runner/src/memory/mod.rs

```diff
@@ -168,8 +174,20 @@ impl Memory {
     /// Returns [`MemoryError::BaseFieldProjectionFailed`] if the value at the address
     /// cannot be projected to a base field element.
     pub fn get_data(&self, addr: M31) -> Result<M31, MemoryError> {
-        let address = addr.0 as usize;
-        let value = self.data.get(address).copied().unwrap_or_default();
+        Self::validate_address(addr)?;
+        let locals_address = addr.0 as usize;
+        let heap_address = MAX_ADDRESS - addr.0 as usize;
```

Comment:
extract in function (repeated)

Reply:
you probably want to have an internal `fn get_qm31_no_trace(&self, addr: M31) -> Result<QM31, MemoryError> {` that you call from the public `get_data` and `get_data_no_trace`